### PR TITLE
Don't use thread_local on mingw or OpenBSD builds

### PR DESCRIPTION
**Pull request by nyalldawson**
MingW has broken support for thread_local, so force disabling it. See:
https://sourceforge.net/p/mingw-w64/bugs/445/
https://sourceforge.net/p/mingw-w64/bugs/527/
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80816

Also OpenBSD has no thread_local support, see https://issues.qgis.org/issues/17351

So on these platforms we fall back to using QThreadStorage.

Fixes #17351

Supersedes #5504

### CODE REVIEW

**Comment by lbartoletti** [comment]
Hi @nyalldawson, please don't merge it now. I will build it today on FreeBSD. I think that we have c++11 thread_local. So, some ifdef will be incorrect.
Thanks.

**Comment by nyalldawson** [comment]
@lbartoletti sure, no rush here.

**Comment by lbartoletti** [comment]
I have build master and #5504 both works on FreeBSD. We have c++11 thread_local (seems using pthread).

File: CMakeLists.txt [hunk 1]
```diff
@@ -142,6 +142,24 @@ IF(WITH_CORE)
     ENDIF (OSGEARTHQT_LIBRARY)
   ENDIF (WITH_GLOBE)
 
+  SET (WITH_THREAD_LOCAL TRUE CACHE BOOL "Determines whether std::thread_local should be used")
+  MARK_AS_ADVANCED(WITH_THREAD_LOCAL)
+
+  IF (MINGW OR CMAKE_SYSTEM_NAME STREQUAL "FreeBSD" OR CMAKE_SYSTEM_NAME STREQUAL "OpenBSD")
```
Comment (lbartoletti) [inline CMakeLists.txt]: you can remove FreeBSD. Thanks.

**Pull request merged**